utils/src/main/kotlin/com/exactpro/th2/lwdataprovider/MessageSearcher.kt

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2023 Exactpro (Exactpro Systems Limited)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exactpro.th2.lwdataprovider
+
+import com.exactpro.th2.common.grpc.Direction
+import com.exactpro.th2.common.grpc.Direction.UNRECOGNIZED
+import com.exactpro.th2.common.message.toJson
+import com.exactpro.th2.common.message.toTimestamp
+import com.exactpro.th2.dataprovider.lw.grpc.DataProviderService
+import com.exactpro.th2.dataprovider.lw.grpc.MessageGroupsSearchRequest
+import com.exactpro.th2.dataprovider.lw.grpc.MessageSearchResponse
+import com.exactpro.th2.dataprovider.lw.grpc.MessageStream
+import com.exactpro.th2.dataprovider.lw.grpc.TimeRelation
+import io.grpc.Context
+import mu.KotlinLogging
+import java.time.Duration
+import java.time.Instant
+import java.util.function.Function
+
+/**
+ * This class provide ability to search messages via lw-data-provider gRPC service
+ * @param searchStep interval for sending requests via [service]. Searcher uses this value to split search interval to smaller pieces to reduce volume of requested data via one time.
+ */
+class MessageSearcher private constructor(
+    private val service: DataProviderService,
+    private val searchStep: Duration = DEFAULT_SEARCH_STEP,
+    private val defaultBook: String,
+    private val defaultSessionGroup: String,
+    private val defaultMessageStreams: Set<MessageStream>,
+) {
+    /**
+     * Searches the first message in [BASE64_FORMAT] format matched by filter. Execute search from [Instant.now] to previous time
+     * @param sessionGroup is parameter for request. Searcher uses [messageStreams] value when [sessionGroup] is null or blank
+     * @param searchInterval for searching. Searcher stops to send requests when the next `from` time is before than search start time minus [searchInterval]
+     */
+    @JvmOverloads
+    fun findLastOrNull(
+        book: String = defaultBook,
+        sessionGroup: String = defaultSessionGroup,
+        messageStreams: Set<MessageStream> = defaultMessageStreams,
+        searchInterval: Duration,
```

Comment:
The required parameter must go first. Otherwise, you cannot use the method without specifying the parameters with default values. But the `filter` lambda can be left at the end